brokerapi/brokers/spanner/broker.go

```diff
@@ -45,182 +43,104 @@ type InstanceInformation struct {
-
-	op, err := db_service.GetCloudOperationByServiceInstanceId(ctx, instanceId)
+	instance, err := db_service.GetServiceInstanceDetailsById(ctx, instanceId)
 	if err != nil {
-		return false, fmt.Errorf("Could not locate CloudOperation in database: %s", err)
-	}
-
-	if _, err := db_service.GetServiceInstanceDetailsById(ctx, instanceId); err != nil {
 		return false, brokerapi.ErrInstanceDoesNotExist
 	}
 
-	// we're polling on instance deletion, which is synchronous, unlike creation. Exit early if the instance has been deleted
-	wasDelete, err := s.LastOperationWasDelete(ctx, instanceId)
-
-	if err != nil {
-		return false, fmt.Errorf("Can't check last operation type: %s", err)
+	if instance.OperationType == models.ClearOperationType {
+		return false, fmt.Errorf("No pending operations could be found for this Spanner instance.")
 	}
-	if wasDelete {
-		return true, nil
+
+	if instance.OperationType != models.ProvisionOperationType {
+		return false, fmt.Errorf("Couldn't poll Spanner instance, unknown operation type: %s", instance.OperationType)
 	}
 
-	ct := option.WithTokenSource(s.HttpConfig.TokenSource(ctx))
-	client, err := googlespanner.NewInstanceAdminClient(ctx, ct)
+	client, err := s.createAdminClient(ctx)
 	if err != nil {
-		return false, fmt.Errorf("Error creating client: %s", err)
+		return false, err
 	}
 
-	spannerOp := client.CreateInstanceOperation(op.Name)
-
-	spannerInstance, err := spannerOp.Poll(ctx)
+	// From https://godoc.org/cloud.google.com/go/spanner/admin/instance/apiv1#CreateInstanceOperation.Poll
+	spannerOp := client.CreateInstanceOperation(instance.OperationId)
+	_, err = spannerOp.Poll(ctx)
 	done := spannerOp.Done()
 
-	// from https://godoc.org/cloud.google.com/go/spanner/admin/instance/apiv1#InstanceOperation.Poll
-	if spannerInstance == nil && err != nil && !done {
+	switch {
+	case err != nil && !done: // There was a failure polling
 		return false, fmt.Errorf("Error checking operation status: %s", err)
-	} else if spannerInstance == nil && err != nil && done {
-		op.Status = "FAILED"
-		op.ErrorMessage = err.Error()
-
-		if dberr := db_service.SaveCloudOperation(ctx, op); dberr != nil {
-			return false, fmt.Errorf(`Error saving operation details to database: %s.`, dberr)
-		}
 
+	case err != nil && done: // The operation completed in error
 		return true, fmt.Errorf("Error provisioning instance: %v", err)
-	} else if spannerInstance == nil && err == nil && !done {
-		op.Status = string(instancepb.Instance_STATE_UNSPECIFIED)
 
-		if err := db_service.SaveCloudOperation(ctx, op); err != nil {
-			return false, fmt.Errorf(`Error saving operation details to database: %s.`, err)
+	case err == nil && done: // The operation was successful
+		instance.OperationId = ""
+		instance.OperationType = models.ClearOperationType
+		if err := db_service.SaveServiceInstanceDetails(ctx, instance); err != nil {
+			s.Logger.Error("updating instance after provision", err)
```

Comment:
This block seems to swallow the error and not return it. That seems contradictory with the design of the rest of the method.